dts: Add _STRING_TOKEN and _STRING_UPPER_TOKEN to string-array

rkel opened the pull request:
This is a proposal of DTS update to allow also accessing string array elements as a tokens.

rkel committed:
dts: Add _STRING_TOKEN and _STRING_UPPER_TOKEN to string-array

This commit adds string token versions of the values also
in items inside string-array.

Signed-off-by: Radosław Koppel <r.koppel@k-el.com>
Co-authored-by: Marti Bolivar <marti.bolivar@nordicsemi.no>
Co-authored-by: Kumar Gala <galak@kernel.org>

galak reviewed (changes requested):
Need some testcases added.  Also asked for some explanation of need/what problem does this solve.

mbolivar-nordic reviewed (commented):
Agree with @galak

rkel commented:
The real need for that is that I am now creating segment LCD driver (sorry, but it has to be closed source at least for now). And for the convenience the description of the segments directly responds how it is described in the documentation. So in every documentation for the LCD we have a segment table and the description then looks like this:

```
com1 =  "S1"  , "0B"  , "1A"  , "S2"  , "2A"  , "MIN" , "3A"  , "3P"  , "4A"  , ""    , "5A"  , "S8"  ;
com2 =  "S6"  , "1F"  , "1B"  , "2F"  , "2B"  , "3F"  , "3B"  , "4F"  , "4B"  , "MAX" , "5B"  , "5F"  ;
com3 =  "S7"  , "1G"  , "1C"  , "2G"  , "2C"  , "3G"  , "3C"  , "4G"  , "4C"  , "S10" , "5C"  , "5G"  ;
com4 =  ""    , "1E"  , "1D"  , "2E"  , "2D"  , "3E"  , "3D"  , "4E"  , "4D"  , "S5"  , "5D"  , "5E"  ;
com-list = < 1 2 3 4 >;
```

Then I have to glue the segments mnemonics together with the prefixes to use them in the driver. I just think that the change here does not break anything in current DT usage and the community can benefit from that change and I was surprised that we do not have it already when we have the same solution for the string value types.

When it comes for the tests - I would be happy to implement it. Just give me some hint - where the test for gen_defines.py is implemented. I would read it and I will propose some tests required.

mbolivar-nordic commented on include/zephyr/devicetree.h:
```diff
@@ -911,6 +913,95 @@
+/**
+ * @brief Get a string-array item value as a token.
+ *
+ * This removes "the quotes" from string-valued item, and converts
+ * non-alphanumeric characters to underscores. That can be useful, for example,
+ * when programmatically using the value to form a C variable or code.
```
```suggestion
 * This removes "the quotes" from an element in the array, and converts
 * non-alphanumeric characters to underscores. That can be useful, for example,
 * when programmatically using the value to form a C variable or code.
```

mbolivar-nordic commented on include/zephyr/devicetree.h:
```diff
@@ -911,6 +913,95 @@
+ *     DT_STRING_TOKEN_BY_IDX(DT_NODELABEL(n2), prop, 0) // 123_foo
+ *     DT_STRING_TOKEN_BY_IDX(DT_NODELABEL(n2), prop, 1) // 456_FOO
+ *
+ * For more notices see @ref DT_STRING_TOKEN.
```
```suggestion
 * For more information, see @ref DT_STRING_TOKEN.
```

mbolivar-nordic replied:
Unresolving this one since there are other cases in this PR where the same change needs to be done. Sorry I was vague. This was meant as an example of what to do in general, so all the cases where "notices" was used before, "information," should be used instead.

mbolivar-nordic commented on include/zephyr/devicetree.h:
```diff
@@ -911,6 +913,95 @@
+ * For more notices see @ref DT_STRING_TOKEN.
+ *
+ * @param node_id node identifier
+ * @param prop lowercase-and-underscores property string name
```
```suggestion
 * @param prop lowercase-and-underscores property name
```

mbolivar-nordic replied:
Unresolving this one since there are other places in the PR (at least the INST versions) where this change should be made as well.

rkel replied:
This was direct copy from other property string macros. I located one more place in my PR that can be fixed this way, but many other places where this sentence is used.

mbolivar-nordic commented on include/zephyr/devicetree.h:
```diff
@@ -911,6 +913,95 @@
+ * @param node_id node identifier
+ * @param prop lowercase-and-underscores property string name
+ * @param idx the index to get
+ * @return the value of @p prop as a token, i.e. without any quotes
+ *         and with special characters converted to underscores
```
```suggestion
 * @return the element in @p prop at index @p idx as a token
```

mbolivar-nordic commented on include/zephyr/devicetree.h:
```diff
@@ -911,6 +913,95 @@
+ *         and with special characters converted to underscores
+ */
+#define DT_STRING_TOKEN_BY_IDX(node_id, prop, idx) \
+	DT_CAT4(node_id, _P_, prop##_IDX_##idx, _STRING_TOKEN)
```
Use DT_CAT6 instead. See the comment above DT_CAT for an explanation of these macros.

rkel replied:
Done

mbolivar-nordic commented on include/zephyr/devicetree.h:
```diff
@@ -911,6 +913,95 @@
+ *         and with special characters converted to underscores
+ */
+#define DT_STRING_UPPER_TOKEN_BY_IDX(node_id, prop, idx) \
+	DT_CAT4(node_id, _P_, prop##_IDX_##idx, _STRING_UPPER_TOKEN)
```
Use DT_CAT6.

rkel replied:
Done

mbolivar-nordic reviewed (changes requested):
Thanks for explaining the use case. I am fine with merging array helpers for this. Here are the remaining blockers from my side:

- documentation issues (see comments; I filed https://github.com/zephyrproject-rtos/zephyr/pull/46344 to improve the other docstrings)
- add test cases for every new API macro (see tests/lib/devicetree/api)
- add instance versions DT_INST_STRING_TOKEN_BY_IDX, DT_INST_STRING_UPPER_TOKEN_BY_IDX, since we also have DT_INST_STRING_TOKEN and DT_INST_STRING_UPPER_TOKEN

mbolivar-nordic commented:
Looks like this needs a rebase also, for the CI issues.

rkel commented:
Rebased

rkel commented:
TBD:

- [ ] Instance versions of the macros
- [ ] Test cases

rkel commented:
I can see that libraries.devicetree.api test cannot fit lpcxpresso55s69_ns. I was not expecting that this test has to fit some physical MCU - do we really need limitation like this?

mbolivar-nordic commented:
> I can see that libraries.devicetree.api test cannot fit lpcxpresso55s69_ns. I was not expecting that this test has to fit some physical MCU - do we really need limitation like this?

No, we don't. Can you figure out the testcase.yaml changes needed to make this just run on some QEMU platforms in upstream CI? I can't ever manage to figure out how twister works here, but I will for sure try to help if you can't figure it out.

rkel committed:
test: devicetree: api: Limit test to native_posix and simulation

This commit limits test targets for the test to run only on native
or simulated machines.

Signed-off-by: Radosław Koppel <r.koppel@k-el.com>

rkel commented:
It seems that forcing it to run on native_posix makes the test skipped. Setting DNM flag until I fix this.

rkel commented:
@mbolivar-nordic I have no idea why it is skipping native_posix configuration. I had tried a few configurations. I think that help from some test team would be appreciated.

mbolivar-nordic commented on tests/lib/devicetree/api/testcase.yaml:
```diff
@@ -3,5 +3,4 @@ tests:
     # will mostly likely be the fastest.
-    integration_platforms:
-      - native_posix
+    platform_allow: native_posix qemu_x86 qemu_x86_64 qemu_cortex_m3
```
Nothing wrong with this line, but the commit message contains this, which should be removed:

```

fix
```

mbolivar-nordic commented on include/zephyr/devicetree.h:
```diff
@@ -50,6 +50,8 @@
  * _IDX_<i>: logical index into property
  * _IDX_<i>_EXISTS: logical index into property is defined
  * _IDX_<i>_PH: phandle array's phandle by index (or phandle, phandles)
+ * _IDX <i>_STRING_TOKEN: string array element value as a token
+ * _IDX <i>_STRING_UPPER_TOKEN: string array element value as a uppercased token
```
Typos here. These should be:

```
 * _IDX_<i>_STRING_TOKEN: string array element value as a token
 * _IDX_<i>_STRING_UPPER_TOKEN: string array element value as a uppercased token
```

Sorry for not noticing earlier.

mbolivar-nordic commented on include/zephyr/devicetree.h:
```diff
@@ -2534,6 +2623,26 @@
 	DT_STRING_UPPER_TOKEN(DT_DRV_INST(inst), prop)
 
+/**
+ * @brief Get an element out of string-arrap property as a token.
```
Get an element out of a string-array property as a token.

mbolivar-nordic replied:
Looks like there is a typo in the latest push: `string-array`, not `string-arrap`

mbolivar-nordic commented on tests/lib/devicetree/api/src/main.c:
```diff
@@ -2315,6 +2315,80 @@ static void test_string_token(void)
+	zassert_equal(DT_STRING_UPPER_TOKEN_BY_IDX(DT_NODELABEL(test_string_array_token_second), val, 3), TOKEN_SECOND_IDX_THREE, "");
+
+	/* Test instances */
+#define STRING_TOKEN_BY_IDX_VAR(node_id) _CONCAT(var_token_, node_id)
```
I think this is quite a clever way to test instance-based APIs without relying on the order of instance numbers. Nice work; this trick hadn't occurred to me.

mbolivar-nordic reviewed (changes requested):
Almost there from my side. Just a few minor things to go. Thanks for your persistence.

mbolivar-nordic commented on include/zephyr/devicetree.h:
```diff
@@ -2534,6 +2623,26 @@
+/**
+ * @brief Get an element out of string-arrap property as a token.
+ * @param inst instance number.
+ * @param prop lowercase-and-unserstores property string name.
```
lowercase-and-underscores property name

(note "unserstores" typo also)

mbolivar-nordic commented on include/zephyr/devicetree.h:
```diff
@@ -2534,6 +2623,26 @@
+/**
+ * @brief Like DT_INST_STRING_TOKEN_BY_IDX(), but uppercased.
+ * @param inst instance number.
+ * @param prop lowercase-and-unserstores property name.
```
same here

mbolivar-nordic commented on include/zephyr/devicetree.h:
```diff
@@ -2534,6 +2623,26 @@
 
+/**
+ * @brief Get an element out of string-arrap property as a token.
+ * @param inst instance number.
```
Are the periods at the end of each parameter new in the latest push? Maybe I just didn't notice them before. But please remove them for consistency with the other docstrings. Example: `@param inst instance number`, not `@param inst instance number.`

mbolivar-nordic commented on scripts/dts/gen_defines.py:
```diff
@@ -650,7 +650,11 @@ def write_vanilla_props(node):
+                    subprop = edtlib.Property(None, subval, None)
+
                     macro2val[macro + f"_IDX_{i}"] = quote_str(subval)
+                    macro2val[macro + f"_IDX_{i}_STRING_TOKEN"] = subprop.val_as_token
+                    macro2val[macro + f"_IDX_{i}_STRING_UPPER_TOKEN"] = subprop.val_as_token.upper()
```
BTW, you can use str_as_token() now instead of creating a fake property. https://github.com/zephyrproject-rtos/zephyr/commit/4ef8c8e44711f68798ff51b8b283003c68e31fcd

rkel replied:
Please note that there was some change provided without any tests and documentation. I am adding my below, but consider removing the previous and left the _STRING_TOKEN version as provided with the whole process, with the documentation and macros.

mbolivar-nordic reviewed (changes requested):
Last round of review comments from me, then I'm +1 to merge.

rkel commented:
@nashif, @galak, @mbolivar-nordic - It is little frustrating how this PR goes. I know I had like 2 weeks vacation during this review. But is is almost 3 months since this was created and a month since my implementing all the changes requested. Is it going to be stale PR? Is there anything we are missing to pull it in?

galak commented on scripts/dts/gen_defines.py:
```diff
@@ -653,6 +653,8 @@ def write_vanilla_props(node):
                     subval_as_token = edtlib.str_as_token(subval)
                     macro2val[macro + f"_IDX_{i}_TOKEN"] = subval_as_token
                     macro2val[macro + f"_IDX_{i}_UPPER_TOKEN"] = subval_as_token.upper()
+                    macro2val[macro + f"_IDX_{i}_STRING_TOKEN"] = subval_as_token
```
we shouldn't have both `_TOKEN` and `_STRING_TOKEN` that are identical.

rkel replied:
I just left it as I did not want to be rude and remove somebody's work. But we had already discussed it with @mbolivar-nordic offline. And it seems that the one that was here sneaked through review in some totally different PR. It is hard for me to follow it now, as the whole discussion is on other machine. I mentioned that here: https://github.com/zephyrproject-rtos/zephyr/pull/46249#discussion_r930419980

galak reviewed (dismissed):
We need to resolve only having one token name in gen_defines.py

galak commented:
https://github.com/zephyrproject-rtos/zephyr/pull/49365 for using common token names.

galak commented:
@rkel can you add a comment here to cleanup soc/arm/xilinx_zynq7000/common/pinctrl_soc.h to use the new macros.

galak commented:
> @nashif, @galak, @mbolivar-nordic - It is little frustrating how this PR goes. I know I had like 2 weeks vacation during this review. But is is almost 3 months since this was created and a month since my implementing all the changes requested. Is it going to be stale PR? Is there anything we are missing to pull it in?

Sorry slide through the cracks.

rkel commented:
Great. Thank you for the support :). I will be able to go deeper through the comments in 2-3 hours since now.

galak commented:
> Great. Thank you for the support :). I will be able to go deeper through the comments in 2-3 hours since now.

@rkel - talked to @mbolivar-nordic and this looks good.  I'll handle my changes requests as a follow up PR.

rkel commented:
Then I am happy as it is now.
